python/lsst/cbp/coordUtils.py

```diff
@@ -0,0 +1,400 @@
+
+    # compute telescope altitude
+    # Use an equation derived from the following, which is from Euler angles:
+    # base z = -sin(negAlt) pupil x + cos(negAlt) pupil z
+    #        = sin(alt) pupil x + cos(alt) pupil z
+
+    xb, yb, zb = spBase.getVector()
+    xp, yp, zp = spPupil.getVector()
+
+    factor = 1 / math.fsum((xp**2, zp**2))
+    addend1 = xp * zb
+    addend2 = zp * math.sqrt(math.fsum((xp**2, zp**2, -zb**2)))
+    sinAltToTry = [val for val in (
+        factor * (addend1 + addend2),
+        factor * (addend1 - addend2),
+    ) if abs(val) <= 1.0]
+    if not sinAltToTry:
+        raise RuntimeError("No solution found; required altitude out of range")
+
+    zerror = 1
+    goodSinAlt = None
+    for i, sinAlt in enumerate(sinAltToTry):
+        cosAlt = math.cos(math.asin(sinAlt))
+        predzb = sinAlt*xp + cosAlt*zp
+        # print("case", i, "sinAlt=", sinAlt, "predzb=", predzb, "zb=", zb, "delta=", predzb-zb)
+        if abs(predzb - zb) < zerror:
+            zerror = abs(predzb - zb)
+            goodSinAlt = sinAlt
+    if goodSinAlt is None:
+        raise RuntimeError("No solution found")
+    alt = math.asin(goodSinAlt)*radians
+
+    # Consider the spherical triangle connecting the telescope pointing (pupil frame x axis),
+    # the vector, and zenith (the base frame z axis). The length of all sides is known.
+    # Solve for the angle between the sides at the zenith;
+    # that angle is the difference in azimuth between the
+    # telescope pointing and the azimuth of the base vector.
+    # A is the side connecting the vector to the pupil frame x axis
+    # (since 0, 0 is a unit vector pointing along pupil frame x);
+    # B is the side connecting telescope pointing to the zenith
+    # C is the side connecting the vector to the zenith
+    sideA = SpherePoint(0, 0, radians).separation(spPupil).asRadians()
+    sideB = (90*degrees - alt).asRadians()
+    sideC = (90*degrees - spBase[1]).asRadians()
+
+    # sideA can be small or zero so use a half angle formula
+    # sides B and C will always be well away from 0 and 180 degrees
+    semiPerimeter = 0.5*math.fsum((sideA, sideB, sideC))
+    sinHalfAngleA = math.sqrt(math.sin(semiPerimeter - sideB) * math.sin(semiPerimeter - sideC) /
+                              (math.sin(sideB) * math.sin(sideC)))
+    daz = 2*math.asin(sinHalfAngleA)*radians
+    if spPupil[0].wrapCtr() > 0:
+        daz = -daz
+    az = spBase[0] + daz
+    global _RecordError
+    if _RecordError:  # to study sources of numerical imprecision
+        global _ErrorLimitArcsec, _ErrorList
+        sp = SpherePoint(az, alt)
+        vectorBaseRT = convertVectorFromPupilToBase(vectorPupil, sp)
+        errorArcsec = SpherePoint(Vector3d(vectorBaseRT)).separation(
+            SpherePoint(Vector3d(*vectorBase))).asArcseconds()
+        if errorArcsec > _ErrorLimitArcsec:
+            _ErrorList.append((errorArcsec, vectorBase, vectorPupil))
+    return SpherePoint(az, alt)
```

Comment:
Aside from `math.fsum`, which I believe has no `numpy` equivalent, I believe both the `math` versions and `numpy` versions give equally good results for scalars. I prefer `math` methods for scalars because I believe it makes the it clear that the inputs and outputs are scalars.